### PR TITLE
[AMS-370] [AMS-262] New advertisement element type "phone"

### PR DESCRIPTION
**Pull request by xakep139**

### CODE REVIEW

File: src/MigrationTool/ImportService.cs [hunk 1]
```diff
@@ -1018,6 +1019,13 @@ private IElementDescriptor ConvertTemplateLinkToTemplateElementDescriptor(Elemen
                         BeginDate = element.BeginDate.Value,
                         EndDate = element.EndDate.Value
                     };
+
+                case ElementDescriptorType.Phone:
+                    return new PhoneElementValue
+                        {
+                            Raw = element.Text,
+                            Formatted = null
```
Comment (xakep139) [inline src/MigrationTool/ImportService.cs]: We don't fill formatted value initially

File: src/VStore/Templates/TemplateElementValidationErrors.cs [hunk 2]
```diff
@@ -7,14 +7,16 @@ public enum TemplateElementValidationErrors
...
         NegativeMaxSize,
         NegativeMaxLines,
+        NegativeMinSymbols,
         NegativeImageSizeDimension,
         MissingSupportedFileFormats,
         MissingSupportedImageSizes,
         EmptySupportedFileFormats,
         EmptySupportedImageSizes,
         NonUniqueTemplateCode,
-        InvalidMaxSymblosPerWord,
```
Comment (xakep139) [inline src/VStore/Templates/TemplateElementValidationErrors.cs]: Fix typo

File: tests/VStore.UnitTests/VStore.UnitTests.csproj [hunk 3]
```diff
@@ -9,7 +9,7 @@
     <GenerateAssemblyConfigurationAttribute>false</GenerateAssemblyConfigurationAttribute>
     <GenerateAssemblyCompanyAttribute>false</GenerateAssemblyCompanyAttribute>
     <GenerateAssemblyProductAttribute>false</GenerateAssemblyProductAttribute>
-    <RootNamespace>NuClear.VStore.UnitTests</RootNamespace>
```
Comment (xakep139) [inline tests/VStore.UnitTests/VStore.UnitTests.csproj]: Fix default namespace

File: src/VStore/Descriptors/Templates/PhoneElementConstraints.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
+
+namespace NuClear.VStore.Descriptors.Templates
+{
+    public class PhoneElementConstraints : IElementConstraints, IEquatable<PhoneElementConstraints>
```
Comment (denisivan0v) [inline src/VStore/Descriptors/Templates/PhoneElementConstraints.cs]: We mustn't have any constraints for phone elements

File: src/VStore/Objects/ContentValidation/Errors/ElementTextTooShortError.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+﻿using Newtonsoft.Json.Linq;
+
+using NuClear.VStore.Json;
+
+namespace NuClear.VStore.Objects.ContentValidation.Errors
+{
+    public class ElementTextTooShortError : ObjectElementValidationError
```
Comment (denisivan0v) [inline src/VStore/Objects/ContentValidation/Errors/ElementTextTooShortError.cs]: We don't have any phone validation on VStore

File: src/VStore/Objects/ContentValidation/PhoneValidator.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+
+using NuClear.VStore.Descriptors.Objects;
+using NuClear.VStore.Descriptors.Templates;
+using NuClear.VStore.Objects.ContentValidation.Errors;
+
+namespace NuClear.VStore.Objects.ContentValidation
+{
+    public static class PhoneValidator
```
Comment (denisivan0v) [inline src/VStore/Objects/ContentValidation/PhoneValidator.cs]: We don't have any phone validation on VStore

File: src/VStore/Templates/TemplatesManagementService.cs [hunk 7]
```diff
@@ -56,7 +56,8 @@ public IReadOnlyCollection<IElementDescriptor> GetAvailableElementDescriptors()
...
                            new ElementDescriptor(ElementDescriptorType.FasComment, 5, new JObject(), new ConstraintSet(new[] { new ConstraintSetItem(Language.Unspecified, new PlainTextElementConstraints()) })),
                            new ElementDescriptor(ElementDescriptorType.Date, 6, new JObject(), new ConstraintSet(new[] { new ConstraintSetItem(Language.Unspecified, new DateElementConstraints()) })),
-                           new ElementDescriptor(ElementDescriptorType.Link, 7, new JObject(), new ConstraintSet(new[] { new ConstraintSetItem(Language.Unspecified, new LinkElementConstraints()) }))
+                           new ElementDescriptor(ElementDescriptorType.Link, 7, new JObject(), new ConstraintSet(new[] { new ConstraintSetItem(Language.Unspecified, new LinkElementConstraints()) })),
+                           new ElementDescriptor(ElementDescriptorType.Phone, 8, new JObject(), new ConstraintSet(new[] { new ConstraintSetItem(Language.Unspecified, new PhoneElementConstraints()) }))
```
Comment (denisivan0v) [inline src/VStore/Templates/TemplatesManagementService.cs]: new PhoneElementConstraints() -> null?

File: tests/VStore.UnitTests/VStore.UnitTests.csproj (hunk 3 above)
Comment (denisivan0v) [inline tests/VStore.UnitTests/VStore.UnitTests.csproj]: 👍

File: src/MigrationTool/ConverterService.cs [hunk 8]
```diff
@@ -192,12 +194,10 @@ private IEnumerable<ImageSize> ConvertImageDimensionToImageSizes(AdvertisementEl
...
             foreach (var size in sizes)
             {
-                int height;
-                int width;
                 var dims = size.Split('x');
                 if (dims.Length != 2 ||
-                    !int.TryParse(dims[0], out width) ||
-                    !int.TryParse(dims[1], out height))
+                    !int.TryParse(dims[0], out int width) ||
+                    !int.TryParse(dims[1], out int height))
```
Comment (xakep139) [inline src/MigrationTool/ConverterService.cs]: Inline declaration

File: src/VStore/Descriptors/Templates/PhoneElementConstraints.cs (hunk 4 above)
Comment (xakep139) [inline src/VStore/Descriptors/Templates/PhoneElementConstraints.cs]: Done in https://github.com/2gis/nuclear-vstore/pull/44/commits/16a0c459407de4a9a55d2123d914ee5d25d70e66

File: src/VStore/Objects/ContentValidation/Errors/ElementTextTooShortError.cs (hunk 5 above)
Comment (xakep139) [inline src/VStore/Objects/ContentValidation/Errors/ElementTextTooShortError.cs]: Removed in https://github.com/2gis/nuclear-vstore/pull/44/commits/16a0c459407de4a9a55d2123d914ee5d25d70e66

File: src/VStore/Objects/ContentValidation/PhoneValidator.cs (hunk 6 above)
Comment (xakep139) [inline src/VStore/Objects/ContentValidation/PhoneValidator.cs]: Removed in https://github.com/2gis/nuclear-vstore/pull/44/commits/16a0c459407de4a9a55d2123d914ee5d25d70e66

File: src/VStore/Templates/TemplatesManagementService.cs (hunk 7 above)
Comment (xakep139) [inline src/VStore/Templates/TemplatesManagementService.cs]: We should discuss this

Comment (denisivan0v) [inline src/VStore/Templates/TemplatesManagementService.cs]: Discussed and have agreed on this.

File: src/VStore.Host/Logging/Serilog.Appender.cs [hunk 9]
```diff
@@ -27,12 +27,7 @@ public class SerilogAppender : AppenderSkeleton
 
         public SerilogAppender(ILogger logger)
         {
-            if (logger == null)
-            {
-                throw new ArgumentNullException(nameof(logger));
-            }
-
-            _logger = logger;
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
```
Comment (xakep139) [inline src/VStore.Host/Logging/Serilog.Appender.cs]: Elvis operator & throw expression

**Pull request merged**